service/raft/raft_rpc.cc

```diff
@@ -33,10 +33,15 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
 {}
 
 
-template <typename Verb, typename Msg> void
+template <raft_rpc::one_way_kind rpc_kind, typename Verb, typename Msg> void
 raft_rpc::one_way_rpc(sloc loc, raft::server_id id,
         Verb&& verb, Msg&& msg) {
     (void)with_gate(_shutdown_gate, [this, loc = std::move(loc), id, &verb, &msg] () mutable {
+        if (rpc_kind == one_way_kind::request && !_failure_detector->is_alive(id)) {
```

Comment:
Why distinguish between requests and replies? We do not do that for two way RPCs though there are kinds that initiated by leaders only and by followers only.

Reply:
The set of nodes that the failure detector pings is based on the current configuration of the group 0 (see `group0_rpc::on_configuration_change`). The current code in the PR treats nodes outside that set as "dead", i.e. rpcs to those nodes will be dropped. IIUC, after adding the node to the configuration, the leader starts sending append_entries probe message to it; if the follower dropped the append_entries_reply rpc there would be no progress made in appending messages to the new node and it wouldn't even learn about the current configuration.

Perhaps dropping the `rpc_kind == one_way_kind::request` and treating nodes outside the raft configuration unconditionally as alive would solve this, but it doesn't feel right.

Reply:
I see. IIRC raft does need to communicate with nodes outside of its config sometimes. One of the cases you describe here: a node that is joining may need to send a reply to a leader before it actually joins. I think there were others, but do not remember from the top of my head. @kostja do you?

Reply:
Leader calculates configuration based on the last appended entry, not the last committed entry. So if the leader appended an entry that adds node to configuration, from leader's POV that node is already in the configuration, so it's not accurate to say "we're sending to a node outside config". The node is inside config at that point, at least from leader's POV.

The problem with responses is, the responding node doesn't yet know that it's part of config.

The question arises: could there be a situation where node X doesn't know about node Y being in config, but needs to send a **request** to it? I don't think there is.

Then there's the issue of joint configurations. But the failure detector is adding nodes that are in the whole joint set, so we should be good.

Reply:
> I see. IIRC raft does need to communicate with nodes outside of its config sometimes. One of the cases you describe here: a node that is joining may need to send a reply to a leader before it actually joins. I think there were others, but do not remember from the top of my head. @kostja do you?

That's the only "legally binding" case, but please keep in mind that old members of configuration can continue communicating with the leader and in this case it is still nice manners to respond and not drop all the incoming messages.

Reply:
Honestly I don't see the point why we need to have this hulk at all. The core matter of the patch is adding server::_is_alive (or rather I would call it _is_started, but oh well), what is the reason to change this piece?

Reply:
Suggestion: drop patch 4 in this series, it is not clear how it is addressing the issue (it should have no impact at all), and it is being a bit too smart. If I decide to send an RPC to a node, I have a reason for it, no need to drop it because hte failure detector thinks so. This pessimizes the algorithm making the node unresponsive unless it is seen as such by the p inger.

Reply:
> Leader calculates configuration based on the last appended entry, not the last committed entry. So if the leader appended an entry that adds node to configuration, from leader's POV that node is already in the configuration, so it's not accurate to say "we're sending to a node outside config". The node is inside config at that point, at least from leader's POV.
> 
> The problem with responses is, the responding node doesn't yet know that it's part of config.
> 
> The question arises: could there be a situation where node X doesn't know about node Y being in config, but needs to send a **request** to it? I don't think there is.
> 
> Then there's the issue of joint configurations. But the failure detector is adding nodes that are in the whole joint set, so we should be good.

The node that is not in the config that we still may need to communicate with is a leader that was removed from the config and we need to tell it we committed the non-joint with him leaving, so now it's time to step down.

Reply:
> Suggestion: drop patch 4 in this series, it is not clear how it is addressing the issue (it should have no impact at all), and it is being a bit too smart. If I decide to send an RPC to a node, I have a reason for it, no need to drop it because hte failure detector thinks so. This pessimizes the algorithm making the node unresponsive unless it is seen as such by the p inger.

The purpose of the series is the patch 4. It is totally fine to trust your failure detector otherwise why do you even have it. We trust it in the core of the raft after all.

Reply:
@kostja if you drop patch 4 issues  https://github.com/scylladb/scylladb/issues/15853 and https://github.com/scylladb/scylladb/issues/15167 will not be fixed, which was the whole point of the series. Please read the cover letter.

Reply:
> Then there's the issue of joint configurations. But the failure detector is adding nodes that are in the whole joint set, so we should be good.

Wait, no, it's only adding the current configuration:
```
    if (_observed._last_conf_idx != _log.last_conf_idx() ||
            (_observed._current_term != _log.last_term() &&
             _observed._last_term != _log.last_term())) {
        configuration last_log_conf = _log.get_configuration();
        last_log_conf.current.merge(last_log_conf.previous);
        output.configuration = last_log_conf.current;
    }
```

When servers are being added this is fine, because `current > previous`, but what for servers that we're removing? If a server is only in `previous` we should still allow sending RPCs to it, but looks like with this patch we wouldn't

Reply:
> The node that is not in the config that we still may need to communicate with is a leader that was removed from the config and we need to tell it we committed the non-joint with him leaving, so now it's time to step down.

But we only need to send responses to it, because that leader is the one who commits the non-joint entry. Unless we're talking about a scenario where there are two leaders running concurrently, but that meants there was some weird failure and we don't need eager stepdown to work in this case, the old leader will eventually stepdown anyway or get killed.

IOW this PR should not cause problems with stepdown. @piodul perhaps you should do a manual test where you decommission the group 0 leader (in non-raft-topology mode) or the topology coordinator (in raft-topology mode), and see if this PR affects the behavior in any way.

Reply:
I just tried this test:

```python
@pytest.mark.asyncio
async def test_decommission(request, manager: ManagerClient):
    logger.info("Bootstrapping the cluster")
    servers = [await manager.server_add(), await manager.server_add(), await manager.server_add()]

    logger.info(f"Decommissioning node {servers[0]}")
    await manager.decommission_node(servers[0].server_id)
```

I turned on trace level logging for the `raft` and `raft_rpc` logger, and it looks like the messages that are printed while the leader steps down are practically the same as if I run the test without the PR.